perun-core/src/main/java/cz/metacentrum/perun/core/impl/Auditer.java

```diff
@@ -791,6 +777,71 @@ public void initialize() throws InternalErrorException {
+	 * returns the list of all resolved messages.
+	 *
+	 * This method is recursive.
+	 *
+	 * When cycle is detected, method will end the recursion and return already generated resolved messages. Also inform about
+	 * this state to the error log. In this case list of resolved messages can be incomplete.
+	 *
+	 * @param session perun session
+	 * @param messages input messages which can cause generating of new resolving messages by registered attr modules
+	 * @param alreadyResolvedMessages list of all already generated messages (used for checking a cycle between two or more registered modules)
+	 *
+	 * @return list of all resolved messages generated by registered attr modules
+	 */
+	private List<String> checkRegisteredAttributesModules(PerunSession session, List<String> messages, List<String> alreadyResolvedMessages) {
+		List<String> addedResolvedMessages = new ArrayList<>();
+		for(String message: messages) {
+			for(VirtualAttributesModuleImplApi virtAttrModuleImplApi : registeredAttributesModules) {
+				log.info("Message {} is given to module {}", message, virtAttrModuleImplApi.getClass().getSimpleName());
+
+				try {
+					addedResolvedMessages.addAll(virtAttrModuleImplApi.resolveVirtualAttributeValueChange((PerunSessionImpl) session, message));
+				} catch (InternalErrorException ex) {
+					log.error("Error when auditer trying to resolve messages in modules.", ex);
+				} catch (WrongAttributeAssignmentException ex) {
+					log.error("Error when auditer trying to resolve messages in modules.", ex);
+				} catch (WrongReferenceAttributeValueException ex) {
+					log.error("Error when auditer trying to resolve messages in modules.", ex);
+				} catch (AttributeNotExistsException ex) {
+					log.error("Error when auditer trying to resolve messages in modules.", ex);
+				} catch (Exception ex) {
+					log.error("An unexpected exception happened when trying to resolve message: {} in module {}, exception: {}",
+							message, virtAttrModuleImplApi.getAttributeDefinition().getFriendlyName(), ex);
+				}
+			}
+		}
+
+		//We still have new resolving messages, so we need to detect if there isn't cycle and if not, continue the recursion
+		if(!addedResolvedMessages.isEmpty()) {
+			//Cycle detection
+			for(String addedResolvedMessage: addedResolvedMessages) {
+				//if message is already present in the list of resolving messages, we are probably in the cycle between two or more registered modules
+				if(alreadyResolvedMessages.contains(addedResolvedMessage)) {
+					//do not continue, just return already resolved messages and inform in the log about this unexpected situation
+					log.error("There is a cycle for resolving message {}. Recursion was stopped forcibly and more resolving messages was not generated! " +
+							"For this reason, some of them may be missing!", addedResolvedMessage);
+					return alreadyResolvedMessages;
```

Comment:
I would only remove repeating message from addedResolvedMessages and continue as usual.

Reply:
You are right, this is much better behavior. I changed that.